src/gencost/waterfall.py

```diff
@@ -2256,3 +2267,480 @@ def valid_generation(df_):
+                x["age_in_current_year"], bins=bins, labels=labels
+            ),
+            ba_plus_age=lambda x: x["ba_plus_essentials"]
+            + "_"
+            + x["age_range"].astype(str),
+        )
+
+    def fill_in_ep_data(self):
+        xwalk = self.xwalk
+
+        hist_data = (
+            self.get_exa_by_generator()
+            .merge(
+                xwalk[["plant_id_eia", "generator_id", "prime_mover", "fuel_group"]],
+                on=["plant_id_eia", "generator_id", "prime_mover"],
+                how="left",
+            )
+            .assign(year=lambda x: x["report_date"].dt.year)
+            .pipe(self.create_fill_in_ep_thresholds)
+        )
+
+        missing_data = (
+            self.find_missing_data()
+            .pipe(self.create_fill_in_ep_thresholds)
+            .assign(
+                essentials_present=lambda x: np.where(
+                    x["essentials"].isin(hist_data["essentials"]), 1, 0
+                ),
+                ba_plus_essentials_present=lambda x: np.where(
+                    x["ba_plus_essentials"].isin(hist_data["ba_plus_essentials"]),
+                    1,
+                    0,
+                ),
+                ba_plus_age_present=lambda x: np.where(
+                    x["ba_plus_age"].isin(hist_data["ba_plus_age"]),
+                    1,
+                    0,
+                ),
+            )  # make an exception of when there is no ba code or age
+            .assign(
+                ba_plus_age_present=lambda x: np.where(
+                    (x["final_ba_code"].isnull())
+                    | (x["final_ba_code"].isna())
+                    | (x["age_in_current_year"].isna())
+                    | (x["age_in_current_year"].isnull()),
+                    0,
+                    x["ba_plus_age_present"],
+                ),
+                ba_plus_essentials_present=lambda x: np.where(
+                    (x["final_ba_code"].isnull()) | (x["final_ba_code"].isna()),
+                    0,
+                    x["ba_plus_essentials_present"],
+                ),
+                fill_in_score=lambda x: x["essentials_present"]
+                + x["ba_plus_essentials_present"]
+                + x["ba_plus_age_present"],
+            )
+        )
+
+        # some sort of for loop that sets prioritizes merging based on fill in score
+
+        three = (
+            missing_data.query("fill_in_score ==3")
+            .merge(
+                hist_data.drop(
+                    columns=[
+                        "plant_id_eia",
+                        "generator_id",
+                        "year",
+                        "prime_mover",
+                        "fuel_group",
+                        "final_ba_code",
+                        "age_in_current_year",
+                        "age_range",
+                        "essentials",
+                        "ba_plus_essentials",
+                    ]
```

Comment:
We're deleting mostly the same columns in each other different score levels, could all that be done at the end after they are concated together?